frontend/src/components/TradeBox/EncryptedChat/EncryptedTurtleChat/index.tsx

```diff
@@ -0,0 +1,304 @@
+        encrypted_private_key: ownEncPrivKey,
+        passphrase: token,
+      },
+      messages: messages,
+    };
+  };
+
+  const onMessage: (dataFromServer: ServerMessage) => void = (dataFromServer) => {
+    if (dataFromServer) {
+      // If we receive an encrypted message
+      if (dataFromServer.message.substring(0, 27) == `-----BEGIN PGP MESSAGE-----`) {
+        decryptMessage(
+          dataFromServer.message.split('\\').join('\n'),
+          dataFromServer.nick == userNick ? ownPubKey : peerPubKey,
+          ownEncPrivKey,
+          token,
+        ).then((decryptedData) => {
+          setLastSent(decryptedData.decryptedMessage === lastSent ? '----BLANK----' : lastSent);
+          setLastIndex(lastIndex < dataFromServer.index ? dataFromServer.index : lastIndex);
+          setMessages((prev: EncryptedChatMessage[]) => {
+            const existingMessage = prev.find((item) => item.index === dataFromServer.index);
+            if (existingMessage) {
+              return prev;
+            } else {
+              return [
+                ...prev,
+                {
+                  index: dataFromServer.index,
+                  encryptedMessage: dataFromServer.message.split('\\').join('\n'),
+                  plainTextMessage: decryptedData.decryptedMessage,
+                  validSignature: decryptedData.validSignature,
+                  userNick: dataFromServer.nick,
+                  time: dataFromServer.time,
+                } as EncryptedChatMessage,
+              ].sort((a, b) => a.index - b.index);
+            }
+          });
+        });
+      }
+      // We allow plaintext communication. The user must write # to start
+      // If we receive an plaintext message
+      else if (dataFromServer.message.substring(0, 1) == '#') {
+        setMessages((prev) => {
+          const existingMessage = prev.find(
+            (item) => item.plainTextMessage === dataFromServer.message,
+          );
+          if (existingMessage) {
+            return prev;
+          } else {
+            return [
+              ...prev,
+              {
+                index: prev.length + 0.001,
+                encryptedMessage: dataFromServer.message,
+                plainTextMessage: dataFromServer.message,
+                validSignature: false,
+                userNick: dataFromServer.nick,
+                time: new Date().toString(),
+              } as EncryptedChatMessage,
+            ].sort((a, b) => a.index - b.index);
+          }
+        });
+      }
+    }
+  };
+
+  const onButtonClicked = (e: any) => {
+    if (token && value.indexOf(token) !== -1) {
+      alert(
+        `Aye! You just sent your own robot token to your peer in chat, that's a catastrophic idea! So bad your message was blocked.`,
+      );
+      setValue('');
+    }
+    // If input string contains '#' send unencrypted and unlogged message
+    else if (value.substring(0, 1) == '#') {
+      apiClient
+        .post(`/api/chat`, {
```

Comment:
If we give it the offset of the last message here, the endpoint replies with the same message we sent. We can use `.then(() ...` to add it to the list of messages in addition setWaitingEcho(false) and setValue: that would make the messages sent feel way less turtle and avoid user sending the same message twice. Would feel similar to Websockets latency for sent messages.

Reply:
Alright I'l add it, maybe you didn't see it but `setWaitingEcho` and `setValue` are already being called on `.finally(...`

Reply:
I saw it! it's what made me think "Oh, the spinner goes away but the new message still takes a few seconds to show on the screen"